[ONNX] Migrate to PT2 logging

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106741
* __->__ #106592

Summary
- The 'dynamo_export' diagnostics leverages the PT2 artifact logger to handle the verbosity
level of logs that are recorded in each SARIF log diagnostic. In addition to SARIF log,
terminal logging is by default disabled. Terminal logging can be activated by setting
the environment variable `TORCH_LOGS="onnx_diagnostics"`. When the environment variable
is set, it also fixes logging level to `logging.DEBUG`, overriding the verbosity level
specified in the diagnostic options.
See `torch/_logging/__init__.py` for more on PT2 logging.
- Replaces 'with_additional_message' with 'Logger.log' like apis.
- Introduce 'LazyString', adopted from 'torch._dynamo.utils', to skip
evaluation if the message will not be logged into diagnostic.
- Introduce 'log_source_exception' for easier exception logging.
- Introduce 'log_section' for easier markdown title logging.
- Updated all existing code to use new api.
- Removed 'arg_format_too_verbose' diagnostic.
- Rename legacy diagnostic classes for TorchScript Onnx Exporter to avoid
confusion.

Follow ups
- The 'dynamo_export' diagnostic now will not capture python stack
information at point of diagnostic creation. This will be added back in
follow up PRs for debug level logging.
- There is type mismatch due to subclassing 'Diagnostic' and 'DiagnosticContext'
for 'dynamo_export' to incorporate with PT2 logging. Follow up PR will
attempt to fix it.
- More docstrings with examples.

BowenBao committed:
[ONNX] Migrate to PT2 logging

Summary
- 'dynamo_export' diagnostics now utilize PT2 artifact logger to manage
the verbosity level of logs recorded into each diagnostic in SARIF log.
The terminal logging is by default turned off. Setting environment variable
'TORCH_LOGS="+onnx_diagnostics"' turns on terminal logging, as well as
adjusts the verbosity level which overrides that in diagnostic options.
Replaces 'with_additional_message' with 'Logger.log' like apis.
- Introduce 'LazyString', adopted from 'torch._dynamo.utils', to skip
evaluation if the message will not be logged into diagnostic.
- Introduce 'log_source_exception' for easier exception logging.
- Introduce 'log_section' for easier markdown title logging.
- Updated all existing code to use new api.
- Removed 'arg_format_too_verbose' diagnostic.
- Rename legacy diagnostic classes for TorchScript Onnx Exporter to avoid
confusion.

Follow ups
- The 'dynamo_export' diagnostic now will not capture python stack
information at point of diagnostic creation. This will be added back in
follow up PRs for debug level logging.
- There is type mismatch due to subclassing 'Diagnostic' and 'DiagnosticContext'
for 'dynamo_export' to incorporate with PT2 logging. Follow up PR will
attempt to fix it.
- More docstrings with examples.

[ghstack-poisoned]

BowenBao committed:
fix ci on "[ONNX] Migrate to PT2 logging"


Summary
- 'dynamo_export' diagnostics now utilize PT2 artifact logger to manage
the verbosity level of logs recorded into each diagnostic in SARIF log.
The terminal logging is by default turned off. Setting environment variable
'TORCH_LOGS="+onnx_diagnostics"' turns on terminal logging, as well as
adjusts the verbosity level which overrides that in diagnostic options.
- Replaces 'with_additional_message' with 'Logger.log' like apis.
- Introduce 'LazyString', adopted from 'torch._dynamo.utils', to skip
evaluation if the message will not be logged into diagnostic.
- Introduce 'log_source_exception' for easier exception logging.
- Introduce 'log_section' for easier markdown title logging.
- Updated all existing code to use new api.
- Removed 'arg_format_too_verbose' diagnostic.
- Rename legacy diagnostic classes for TorchScript Onnx Exporter to avoid
confusion.

Follow ups
- The 'dynamo_export' diagnostic now will not capture python stack
information at point of diagnostic creation. This will be added back in
follow up PRs for debug level logging.
- There is type mismatch due to subclassing 'Diagnostic' and 'DiagnosticContext'
for 'dynamo_export' to incorporate with PT2 logging. Follow up PR will
attempt to fix it.
- More docstrings with examples.

[ghstack-poisoned]

BowenBao committed:
fix ci on "[ONNX] Migrate to PT2 logging"


Summary
- 'dynamo_export' diagnostics now utilize PT2 artifact logger to manage
the verbosity level of logs recorded into each diagnostic in SARIF log.
The terminal logging is by default turned off. Setting environment variable
'TORCH_LOGS="+onnx_diagnostics"' turns on terminal logging, as well as
adjusts the verbosity level which overrides that in diagnostic options.
- Replaces 'with_additional_message' with 'Logger.log' like apis.
- Introduce 'LazyString', adopted from 'torch._dynamo.utils', to skip
evaluation if the message will not be logged into diagnostic.
- Introduce 'log_source_exception' for easier exception logging.
- Introduce 'log_section' for easier markdown title logging.
- Updated all existing code to use new api.
- Removed 'arg_format_too_verbose' diagnostic.
- Rename legacy diagnostic classes for TorchScript Onnx Exporter to avoid
confusion.

Follow ups
- The 'dynamo_export' diagnostic now will not capture python stack
information at point of diagnostic creation. This will be added back in
follow up PRs for debug level logging.
- There is type mismatch due to subclassing 'Diagnostic' and 'DiagnosticContext'
for 'dynamo_export' to incorporate with PT2 logging. Follow up PR will
attempt to fix it.
- More docstrings with examples.

[ghstack-poisoned]

BowenBao committed:
fix ci, add some docs on "[ONNX] Migrate to PT2 logging"


Summary
- The 'dynamo_export' diagnostics leverages the PT2 artifact logger to handle the verbosity
level of logs that are recorded in each SARIF log diagnostic. In addition to SARIF log,
terminal logging is by default disabled. Terminal logging can be activated by setting
the environment variable `TORCH_LOGS="onnx_diagnostics"`. When the environment variable
is set, it also fixes logging level to `logging.DEBUG`, overriding the verbosity level
specified in the diagnostic options.
See `torch/_logging/__init__.py` for more on PT2 logging.
- Replaces 'with_additional_message' with 'Logger.log' like apis.
- Introduce 'LazyString', adopted from 'torch._dynamo.utils', to skip
evaluation if the message will not be logged into diagnostic.
- Introduce 'log_source_exception' for easier exception logging.
- Introduce 'log_section' for easier markdown title logging.
- Updated all existing code to use new api.
- Removed 'arg_format_too_verbose' diagnostic.
- Rename legacy diagnostic classes for TorchScript Onnx Exporter to avoid
confusion.

Follow ups
- The 'dynamo_export' diagnostic now will not capture python stack
information at point of diagnostic creation. This will be added back in
follow up PRs for debug level logging.
- There is type mismatch due to subclassing 'Diagnostic' and 'DiagnosticContext'
for 'dynamo_export' to incorporate with PT2 logging. Follow up PR will
attempt to fix it.
- More docstrings with examples.

[ghstack-poisoned]

BowenBao committed:
add unittests on "[ONNX] Migrate to PT2 logging"


Summary
- The 'dynamo_export' diagnostics leverages the PT2 artifact logger to handle the verbosity
level of logs that are recorded in each SARIF log diagnostic. In addition to SARIF log,
terminal logging is by default disabled. Terminal logging can be activated by setting
the environment variable `TORCH_LOGS="onnx_diagnostics"`. When the environment variable
is set, it also fixes logging level to `logging.DEBUG`, overriding the verbosity level
specified in the diagnostic options.
See `torch/_logging/__init__.py` for more on PT2 logging.
- Replaces 'with_additional_message' with 'Logger.log' like apis.
- Introduce 'LazyString', adopted from 'torch._dynamo.utils', to skip
evaluation if the message will not be logged into diagnostic.
- Introduce 'log_source_exception' for easier exception logging.
- Introduce 'log_section' for easier markdown title logging.
- Updated all existing code to use new api.
- Removed 'arg_format_too_verbose' diagnostic.
- Rename legacy diagnostic classes for TorchScript Onnx Exporter to avoid
confusion.

Follow ups
- The 'dynamo_export' diagnostic now will not capture python stack
information at point of diagnostic creation. This will be added back in
follow up PRs for debug level logging.
- There is type mismatch due to subclassing 'Diagnostic' and 'DiagnosticContext'
for 'dynamo_export' to incorporate with PT2 logging. Follow up PR will
attempt to fix it.
- More docstrings with examples.

[ghstack-poisoned]

thiagocrepaldi commented on torch/onnx/_internal/diagnostics/infra/context.py:
```diff
@@ -96,13 +104,87 @@ def with_graph(self: _Diagnostic, graph: infra.Graph) -> _Diagnostic:
+            self.additional_messages.append(message % args)
+            self.additional_messages.append("\n")
+
+    def debug(self, message: str, *args, **kwargs) -> None:
+        """Logs a debug message within the diagnostic. Same api as logging.Logger.debug.
+
+        Checkout `log` for more details.
+        """
+        self.log(logging.DEBUG, message, *args, **kwargs)
+
+    def info(self, message: str, *args, **kwargs) -> None:
+        """Logs an info message within the diagnostic. Same api as logging.Logger.info.
+
+        Checkout `log` for more details.
+        """
+        self.log(logging.INFO, message, *args, **kwargs)
+
+    def warning(self, message: str, *args, **kwargs) -> None:
+        """Logs a warning message within the diagnostic. Same api as logging.Logger.warning.
+
+        Checkout `log` for more details.
+        """
+        self.log(logging.WARNING, message, *args, **kwargs)
+
+    def error(self, message: str, *args, **kwargs) -> None:
```
Maybe add tests for these APIs to prevent regressions? I think I only saw tests for the `.log` variation

thiagocrepaldi commented on torch/onnx/_internal/diagnostics/infra/context.py:
```diff
@@ -96,13 +104,87 @@ def with_graph(self: _Diagnostic, graph: infra.Graph) -> _Diagnostic:
-            self.additional_message = f"{self.additional_message}\n{message}"
-        return self
+    @contextlib.contextmanager
+    def log_section(
```
test for this?

thiagocrepaldi commented on torch/onnx/_internal/diagnostics/infra/context.py:
```diff
@@ -96,13 +104,87 @@ def with_graph(self: _Diagnostic, graph: infra.Graph) -> _Diagnostic:
-            self.additional_message = f"{self.additional_message}\n{message}"
-        return self
+    @contextlib.contextmanager
+    def log_section(
```
```suggestion
    def log_context(
```
more aligned with the "context manager" concept in python

What is the difference between `diagnostic_context` and `log_section`? Could one be part of the other and publicly the user only care about the only public API?

BowenBao replied:
this is not what it meant. I'll add examples.

thiagocrepaldi commented on torch/onnx/_internal/diagnostics/infra/formatter.py:
```diff
@@ -19,6 +20,35 @@
 ]
 
 
+class LazyString:
```
For my knowledge, if Diagnostic wouldn't be enabled by default, do we still need lazy evaluation? Could we hide the "laziness" within the implementation so we don't have to expose and delegate to user when they should use it when adding more diagnostics on their own?

BowenBao replied:
Diagnostic is enabled by default. 

And on the contrary to your question, we need lazy evaluation more if it is not enabled by default, because that means we have more places where we don't want to evaluate messages since we don't log it.

There is no way around it for delayed evaluation. User must pass a function or lambda w/ args. I took this design since it is already there in PT2 logging so we are keeping similar log api surface.

There are no checks for violation. I'm not aware of any simple solution. The only thing I can think of is creating a linter, hence this is low priority.

thiagocrepaldi commented on torch/onnx/_internal/diagnostics/infra/_infra.py:
```diff
@@ -272,5 +273,8 @@ class DiagnosticOptions:
     Options for diagnostic context.
     """
 
+    verbosity_level: int = dataclasses.field(default=logging.INFO)
```
Do we plan to have another log level outside the context of diagnotic or it is one for all exporter?

If we need only one:
* could we call it `"onnx"` instead of `"onnx_diagnostic"` Given this string is passed to `TORCH_LOGS`, the `_diagnostics` seems redundant
* should it live outside of the context of `DiagnosticOptions`?

BowenBao replied:
This is for diagnostic verbosity level, hence under DiagnosticOptions. The public api on ExportOptions can be bikeshedded.

Please checkout [NOTE: `dynamo_export` diagnostics logging] and PT2 logging for log vs artifact log.

"onnx" is the root logger for anything under `torch.onnx`, "onnx_diagnostics" is the artifact logger under "onnx" logger used by diagnostics.

I'm keeping them separate such that

- It is clear when using `TORCH_LOGS="onnx_diagnostics"` it is the diagnostic verbosity level being adjusted.
- "onnx" root logger can be used for tools and testing, places where export is not called and hence no diagnostic around.

BowenBao committed:
Update on "[ONNX] Migrate to PT2 logging"


Summary
- The 'dynamo_export' diagnostics leverages the PT2 artifact logger to handle the verbosity
level of logs that are recorded in each SARIF log diagnostic. In addition to SARIF log,
terminal logging is by default disabled. Terminal logging can be activated by setting
the environment variable `TORCH_LOGS="onnx_diagnostics"`. When the environment variable
is set, it also fixes logging level to `logging.DEBUG`, overriding the verbosity level
specified in the diagnostic options.
See `torch/_logging/__init__.py` for more on PT2 logging.
- Replaces 'with_additional_message' with 'Logger.log' like apis.
- Introduce 'LazyString', adopted from 'torch._dynamo.utils', to skip
evaluation if the message will not be logged into diagnostic.
- Introduce 'log_source_exception' for easier exception logging.
- Introduce 'log_section' for easier markdown title logging.
- Updated all existing code to use new api.
- Removed 'arg_format_too_verbose' diagnostic.
- Rename legacy diagnostic classes for TorchScript Onnx Exporter to avoid
confusion.

Follow ups
- The 'dynamo_export' diagnostic now will not capture python stack
information at point of diagnostic creation. This will be added back in
follow up PRs for debug level logging.
- There is type mismatch due to subclassing 'Diagnostic' and 'DiagnosticContext'
for 'dynamo_export' to incorporate with PT2 logging. Follow up PR will
attempt to fix it.
- More docstrings with examples.

[ghstack-poisoned]

titaiwangms commented:
Can https://github.com/pytorch/pytorch/pull/106478 be merged before this PR lol

BowenBao commented:
> Can #106478 be merged before this PR lol

Sure, yours is earlier :)

titaiwangms commented on torch/onnx/_internal/diagnostics/infra/context.py:
```diff
@@ -96,13 +104,87 @@ def with_graph(self: _Diagnostic, graph: infra.Graph) -> _Diagnostic:
+            self.additional_messages.append(message % args)
+            self.additional_messages.append("\n")
+
+    def debug(self, message: str, *args, **kwargs) -> None:
```
Not sure how this affects internal api. I only see .info being used.

BowenBao replied:
Right now only graph diff in `_pass.py` is logged with debug. Updating docstring

titaiwangms commented on torch/onnx/_internal/diagnostics/infra/formatter.py:
```diff
@@ -19,6 +20,35 @@
 
 
+class LazyString:
+    """A class to lazily evaluate a string.
```
For my education, what does lazily evaluate a string mean?

BowenBao replied:
In short, it allows passing a lambda instead of string to `log` api. So logger, based on log level, decides whether to call the lambda. 

Useful when you have a message that takes a long time to format, you can log with debug level so normal export performance (default level info) is not affected.

titaiwangms commented on torch/onnx/_internal/fx/op_validation.py:
```diff
@@ -76,10 +78,10 @@ def validate_op_between_ort_torch(
+        with diagnostic.log_section(
+            logging.WARNING, "Op level debug fails due to unsupported input types"
+        ):
+            diagnostic.log_source_exception(logging.WARNING, value_error)
```
.warning is not used here, because we need `value_error`?

BowenBao replied:
`logging.WARNING` is passed as argument.

titaiwangms reviewed (commented):
Is diagnostics.levels.ERROR affected?

BowenBao committed:
rebase w/ dispatcher change on "[ONNX] Migrate to PT2 logging"


Summary
- The 'dynamo_export' diagnostics leverages the PT2 artifact logger to handle the verbosity
level of logs that are recorded in each SARIF log diagnostic. In addition to SARIF log,
terminal logging is by default disabled. Terminal logging can be activated by setting
the environment variable `TORCH_LOGS="onnx_diagnostics"`. When the environment variable
is set, it also fixes logging level to `logging.DEBUG`, overriding the verbosity level
specified in the diagnostic options.
See `torch/_logging/__init__.py` for more on PT2 logging.
- Replaces 'with_additional_message' with 'Logger.log' like apis.
- Introduce 'LazyString', adopted from 'torch._dynamo.utils', to skip
evaluation if the message will not be logged into diagnostic.
- Introduce 'log_source_exception' for easier exception logging.
- Introduce 'log_section' for easier markdown title logging.
- Updated all existing code to use new api.
- Removed 'arg_format_too_verbose' diagnostic.
- Rename legacy diagnostic classes for TorchScript Onnx Exporter to avoid
confusion.

Follow ups
- The 'dynamo_export' diagnostic now will not capture python stack
information at point of diagnostic creation. This will be added back in
follow up PRs for debug level logging.
- There is type mismatch due to subclassing 'Diagnostic' and 'DiagnosticContext'
for 'dynamo_export' to incorporate with PT2 logging. Follow up PR will
attempt to fix it.
- More docstrings with examples.

[ghstack-poisoned]

BowenBao commented on docs/source/onnx_diagnostics.rst:
```diff
@@ -28,5 +28,5 @@ Diagnostic Rules
 -------------
 
-.. autoclass:: torch.onnx._internal.diagnostics.ExportDiagnostic
+.. autoclass:: torch.onnx._internal.diagnostics.TorchScriptOnnxExportDiagnostic
```
More to separate new one from old one.

In doc updating follow ups, these torchscript API references should be removed.

BowenBao commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -1509,8 +1519,8 @@ def _maybe_promote_arg(
-                diagnostic.with_additional_message(
-                    f"Argument {fx_arg} is not promoted. Already {dtype}."
+                diagnostic.info(
+                    "Argument %s is not promoted. Already %s.", fx_arg, dtype
```
fstring is evaluated even if not logged which hinders performance.

BowenBao commented on torch/onnx/_internal/fx/op_validation.py:
```diff
@@ -128,11 +126,8 @@ def validate_op_between_ort_torch(
-            diagnostic.with_source_exception(type_error)
+            with diagnostic.log_section(logging.WARNING, "Op level debug is bypassed"):
+                diagnostic.log_source_exception(logging.WARNING, type_error)
             diagnostic.level = diagnostics.levels.WARNING
```
For now diagnostic level and log level are treated separately. Let me know if you have suggestions.

BowenBao commented on test/onnx/internal/test_diagnostics.py:
```diff
@@ -12,8 +13,9 @@
 from torch.onnx import errors
 from torch.onnx._internal import diagnostics
 from torch.onnx._internal.diagnostics import infra
```
yea, all diagnostics tests are in this same file. The infra.DiagnosticContext is for unittests on base diagnostics infra alone, which does not involve any torch or onnx concept.

BowenBao committed:
Fix CI, rebase, address comments on "[ONNX] Migrate to PT2 logging"


Summary
- The 'dynamo_export' diagnostics leverages the PT2 artifact logger to handle the verbosity
level of logs that are recorded in each SARIF log diagnostic. In addition to SARIF log,
terminal logging is by default disabled. Terminal logging can be activated by setting
the environment variable `TORCH_LOGS="onnx_diagnostics"`. When the environment variable
is set, it also fixes logging level to `logging.DEBUG`, overriding the verbosity level
specified in the diagnostic options.
See `torch/_logging/__init__.py` for more on PT2 logging.
- Replaces 'with_additional_message' with 'Logger.log' like apis.
- Introduce 'LazyString', adopted from 'torch._dynamo.utils', to skip
evaluation if the message will not be logged into diagnostic.
- Introduce 'log_source_exception' for easier exception logging.
- Introduce 'log_section' for easier markdown title logging.
- Updated all existing code to use new api.
- Removed 'arg_format_too_verbose' diagnostic.
- Rename legacy diagnostic classes for TorchScript Onnx Exporter to avoid
confusion.

Follow ups
- The 'dynamo_export' diagnostic now will not capture python stack
information at point of diagnostic creation. This will be added back in
follow up PRs for debug level logging.
- There is type mismatch due to subclassing 'Diagnostic' and 'DiagnosticContext'
for 'dynamo_export' to incorporate with PT2 logging. Follow up PR will
attempt to fix it.
- More docstrings with examples.

[ghstack-poisoned]

thiagocrepaldi reviewed (commented):
Going OOF, so I will defer this to our colleagues. :)

The only thing I spotted is having a global logger instead of part of Diagnostic classes

BowenBao commented on torch/onnx/_internal/diagnostics/infra/context.py:
```diff
@@ -16,6 +18,7 @@
 
 # This is a workaround for mypy not supporting Self from typing_extensions.
 _Diagnostic = TypeVar("_Diagnostic", bound="Diagnostic")
+diagnostic_logger: logging.Logger = logging.getLogger(__name__)
```
Similar discussion here https://github.com/pytorch/pytorch/pull/106741#discussion_r1291689604 on idiomatic global logger.

It is a bit anti-pattern that Diagnostic needs to keep track of logger, but I feel it is better to keep track of the global logger instead of a global var of `__name__` value here to retrieve that same logger.

BowenBao commented:
@pytorchbot merge

BowenBao commented:
Merging but will address any more comments in follow ups.
